### DISCUSSION TITLE
External TDD PA/LNA using GPIO

### DISCUSSION THREAD

**Discussion by yc541** [srsran/srsRAN_Project#276] [General Help] [answered]
Hello,

I am trying to run an external TDD PA/LNA with a USRP X300 using its front panel GPIO pins, which output high/low signals to toggle PA or LNA. The logic is outputting GPIO high (PA on, LNA off) when the USRP needs to transmit and outputting GPIO low (PA off, LNA on) when the USRP needs to receive. There is a post on the same topic https://github.com/srsran/srsRAN_Project/discussions/43 so I tried the suggested method in there.

I put the code of outputting GPIO high in the code block  _if (md.start_of_burst)_ (SRS version is 23_10 so the variable name is slightly different) suggested by @rainey19, and the code of outputting GPIO low in the code block _if (md.end_of_burst)_.

The result is that the GPIO is always high. I did a bit more reading of the FSM written in _lib/radio/uhd)/radio_uhd_tx_stream_fsm.h_ and found that the FSM only comes into START_BURST state once at initialization and stays in IN_BURST state forever, so the code pretty much never goes into the code blocks _if (md.start_of_burst)_ nor  _if (md.end_of_burst)_.

I also tried putting the GPIO code directly in every call of _radio_uhd_tx_stream::transmit_block_ and _radio_uhd_rx_stream::receive_block_ but this resulted too many late and underflow. 

Could someone point me to the right direction of implementing this? Many thanks in advance.

**Comment by dustinmaas** [comment, marked as answer]
FWIW. My team ran into this issue last week and we have added the necessary end-of-burst messaging. This required plumbing information about the type (UL or DL) for the current slot/symbols from the MAC down to the PHY, as well as making some small timing adjustments to make up for the switching delay. Our patched version of the code is now working for us outdoors with a front end and we plan to do some cleanup and submit a pull request in the near future.

**Reply by yc541** [reply]
Many thanks for the reply and good to hear you have a solution. Look forward to testing your patch later!

**Comment by pgawlowicz** [comment]
what about using timed commands available in the UHD API and scheduling a GPIO change just before TX time?
https://github.com/manuts/uhd-examples/blob/master/test_timed_commands.cpp

Then scheduling another command to toggle the GPIO after TX.

**Comment by joaquim-srs** [comment]
Hi there,

As of last week's release, srsRAN includes a discontinuous transmission mode which may address your needs. At this moment, the feature is experimental, we are still testing it and open to feedback. You can take a look at the [feature announcement discussion](https://github.com/srsran/srsRAN_Project/discussions/367) for more details.
